src/seccomp/src/compiler.rs

```diff
@@ -0,0 +1,561 @@
+
+//! Module responsible for compiling the deserialized filter objects into BPF code.
+//! Used by the seccompiler binary.
+//!
+//! It also defines most of the objects that a seccomp filter is deserialized into:
+//! `Filter`, `SyscallObject`.
+//!
+//! The rest of objects are deserialized directly into the IR (intermediate representation), declared in `lib.rs`:
+//! `SeccompCondition`, `SeccompAction`, `SeccompCmpOp`, `SeccompCmpArgLen`.
+//!
+//! The compilation goes through a couple of steps, from JSON to BPF:
+//!                  JSON
+//!                   |
+//!            (via serde_json)
+//!                   |
+//!                   V
+//!       collection of `Filter` objects
+//!                   |
+//!         (via Compiler.compile_blob(...))
+//!                   |
+//!                   V
+//!   collection of `SeccompFilter` objects
+//!     (IR - intermediate representation)
+//!                   |
+//!    (via SeccompFilter.into_bpf(...))
+//!                   |
+//!                   V
+//!     collection of `BpfProgram` objects
+
+use super::syscall_table::SyscallTable;
+use seccomp::{
+    BpfThreadMap, Error as SeccompFilterError, SeccompAction, SeccompCondition, SeccompFilter,
+    SeccompRule, SeccompRuleMap,
+};
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fmt;
+
+/// Errors compiling Filters into BPF
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Erroneous SyscallObject with both `syscall` and `syscalls`
+    MultipleSyscallFields,
+    /// Error from the SeccompFilter
+    SeccompFilter(SeccompFilterError),
+    /// Invalid syscall name for the given arch
+    SyscallName(String, String),
+    /// Erroneous `syscalls` object with argument checks
+    SyscallsWithArgs(Vec<String>),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            MultipleSyscallFields => write!(
+                f,
+                "Cannot use both `syscalls` and `syscall` properties in an object"
+            ),
+            SeccompFilter(ref err) => write!(f, "{}", err),
+            SyscallName(ref syscall_name, ref arch) => write!(
+                f,
+                "Invalid syscall name: {} for given arch: {}",
+                syscall_name, arch
+            ),
+            SyscallsWithArgs(ref syscalls) => write!(
+                f,
+                "The object with the following `syscalls`: {:?} cannot have argument\
+                 conditions. Use the `syscall` property instead or remove the `args` property.",
+                syscalls
+            ),
+        }
+    }
+}
+
+/// Deserializable object encapsulating rules for one or multiple syscalls
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct SyscallObject {
+    /// Name of the syscall
+    syscall: Option<String>,
```

Comment:
Can we drop this field? Just remove the Option from syscalls and enforce the user to always specify at least one.

Reply:
I actually proposed this in the initial design of the file format. However, during the design reviews, the general opinion was that we should have two distinct types of objects: 

- one that refers only to a single syscall and can have argument checks.
- anther that refers to a collection of syscalls and cannot have arguments conditions.

This arguably makes for a easier to use API, and I think this is one of the top priorities. WDYT?

Reply:
done